File: protocol/platform/descriptor/descriptor.go
// Package descriptor parses and validates the platform descriptor, the single
// source of truth for a robot's shape (schema: protocol/platform/SCHEMA.md).
package descriptor

import (
	"fmt"
	"os"

	"github.com/BurntSushi/toml"
)

type Descriptor struct {
	Schema       int               `toml:"schema"`
	Platform     Platform          `toml:"platform"`
	Drivers      map[string]Driver `toml:"drivers"`
	Joints       []Joint           `toml:"joints"`
	Kinematics   *Kinematics       `toml:"kinematics"`
	Allocation   *Allocation       `toml:"allocation"`
	Sensors      []Sensor          `toml:"sensors"`
	Modes        *Modes            `toml:"modes"`
	Observations *Observations     `toml:"observations"`
	Actions      *Actions          `toml:"actions"`
}

type Platform struct {
	ID           string `toml:"id"`
	Name         string `toml:"name"`
	VehicleClass string `toml:"vehicle_class"`
}

type Driver struct {
	Kind string `toml:"kind"`
	Port string `toml:"port"`
	Baud int    `toml:"baud"`
}

type Joint struct {
	Name              string   `toml:"name"`
	Driver            string   `toml:"driver"`
	BusID             uint32   `toml:"bus_id"`
	Type              string   `toml:"type"`
	Ownership         string   `toml:"ownership"`
	Invert            bool     `toml:"invert"`
	CommandInterfaces []string `toml:"command_interfaces"`
	StateInterfaces   []string `toml:"state_interfaces"`
	Limits            *Limits  `toml:"limits"`
}

type Limits struct {
	VelocityRadps  *float64 `toml:"velocity_radps"`
	PositionMinRad *float64 `toml:"position_min_rad"`
	PositionMaxRad *float64 `toml:"position_max_rad"`
}

type Kinematics struct {
	Model        string            `toml:"model"`
	WheelRadiusM float64           `toml:"wheel_radius_m"`
	TrackWidthM  float64           `toml:"track_width_m"`
	Wheels       map[string]string `toml:"wheels"`
}

// Allocation is required only for vehicle classes with an explicit
// effectiveness matrix (none in schema v1; diff_drive derives from kinematics).
type Allocation struct{}

type Sensor struct {
	Name   string  `toml:"name"`
	Kind   string  `toml:"kind"`
	Frame  string  `toml:"frame"`
	RateHz float64 `toml:"rate_hz"`
}

type Modes struct {
	Available []string `toml:"available"`
}

type Observations struct {
	Streams []Stream `toml:"streams"`
}

type Stream struct {
	Subject string  `toml:"subject"`
	Message string  `toml:"message"`
	RateHz  float64 `toml:"rate_hz"`
}

type Actions struct {
	Altitudes []string `toml:"altitudes"`
}

// Parse decodes and validates a descriptor. Unknown keys are errors.
func Parse(data []byte) (*Descriptor, error) {
	var d Descriptor
	md, err := toml.Decode(string(data), &d)
	if err != nil {
		return nil, fmt.Errorf("descriptor: %w", err)
	}
	if un := md.Undecoded(); len(un) > 0 {
		return nil, fmt.Errorf("descriptor: unknown key %q", un[0].String())
	}
	if err := d.validate(); err != nil {
		return nil, err
	}
	return &d, nil
}

// Load reads and parses a descriptor file.
func Load(path string) (*Descriptor, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return Parse(data)
}

File: protocol/platform/ts/descriptor.ts
// Platform descriptor loader for TypeScript consumers. Parsing and derived
// values only; full validation is the Go package's job (the descriptor a
// dashboard sees has already been validated on the rover). Kept in golden-
// fixture conformance with the Go implementation.
import { parse } from 'smol-toml';

export interface Joint {
  name: string;
  driver: string;
  bus_id: number;
  type: 'wheel' | 'revolute' | 'gripper';
  ownership: 'platform' | 'module';
  invert?: boolean;
  command_interfaces: string[];
  state_interfaces: string[];
  limits?: {
    velocity_radps?: number;
    position_min_rad?: number;
    position_max_rad?: number;
  };
}

export interface PlatformDescriptor {
  schema: number;
  platform: { id: string; name?: string; vehicle_class: string };
  drivers: Record<string, { kind: string; port?: string; baud?: number }>;
  joints: Joint[];
  kinematics?: {
    model: string;
    wheel_radius_m: number;
    track_width_m: number;
    wheels: Record<string, string>;
  };
  modes?: { available: string[] };
  observations?: { streams: { subject: string; message: string; rate_hz: number }[] };
  actions?: { altitudes: string[] };
}

const KNOWN_VEHICLE_CLASSES = new Set(['diff_drive_rover', 'fixed_base']);

export function parseDescriptor(toml: string): PlatformDescriptor {
  const d = parse(toml) as unknown as PlatformDescriptor;
  if (d.schema !== 1) {
    throw new Error(`descriptor: unsupported schema ${d.schema}`);
  }
  if (!KNOWN_VEHICLE_CLASSES.has(d.platform.vehicle_class)) {
    throw new Error(`descriptor: unknown vehicle_class ${d.platform.vehicle_class}`);
  }
  if (d.platform.vehicle_class === 'diff_drive_rover' && !d.kinematics) {
    throw new Error('descriptor: diff_drive_rover requires [kinematics]');
  }
  if (d.platform.vehicle_class === 'fixed_base' && d.kinematics) {
    throw new Error('descriptor: fixed_base must not declare [kinematics]');
  }
  return d;
}

export function busIdForJoint(d: PlatformDescriptor, name: string): number | undefined {
  return d.joints.find((j) => j.name === name)?.bus_id;
}

/** Bus ids of platform-owned joints: the module-surface deny-list. */
export function platformOwnedBusIds(d: PlatformDescriptor): number[] {
  return d.joints.filter((j) => j.ownership === 'platform').map((j) => j.bus_id);
}

/**
 * Body twist to signed wheel angular velocities (rad/s) keyed by joint name,
 * inversion applied: the value a driver writes, not the rover-frame value.
 */
export function wheelSpeeds(
  d: PlatformDescriptor,
  vxMps: number,
  wzRadps: number,
): Record<string, number> {
  const k = d.kinematics;
  if (!k || k.model !== 'diff_drive') {
    throw new Error('descriptor: wheelSpeeds requires diff_drive kinematics');
  }
  const vL = vxMps - (wzRadps * k.track_width_m) / 2;
  const vR = vxMps + (wzRadps * k.track_width_m) / 2;
  const out: Record<string, number> = {};
  for (const [pos, name] of Object.entries(k.wheels)) {
    const v = pos === 'front_right' || pos === 'back_right' ? vR : vL;
    let w = v / k.wheel_radius_m;
    if (d.joints.find((j) => j.name === name)?.invert) w = -w;
    out[name] = w;
  }
  return out;
}
